Broadcasting feature

DeadMor0z opened the pull request:
Show markers from nearest and/or biggest range broadcasting ore detectors - grid with ore detector shoud have working and broadcasting antenna (it's radius will be used to check if player in range) and working ore detector set to broadcast markers.
If several grids found with broadcasting ore detector - one with biggest area coverage relative to player position will be used.
Only those grids will be taken into account where player has access to ore detector and antenna (i.e. playe or faction owned)

Commit:
Show markers from nearest and (or) biggest range broadcasting ore detector

Racher commented:
I will look into it but here is an issue I dont think can be perfectly solved without a performance hit:
(didnt look into your code yet but based on your description you didnt solve it either)
Lets say you are on the edge of 2 equal range detectors and there is an iron next to you within range of only one of them. I dont want it to flicker(or show another random iron 5km away) based on which detector is barely closer.

A possible compromise is to use the player as the center of the search with range of (detector range - detector distance).

DeadMor0z commented:
> I will look into it but here is an issue I dont think can be perfectly solved without a performance hit: (didnt look into your code yet but based on your description you didnt solve it either) Lets say you are on the edge of 2 equal range detectors and there is an iron next to you within range of only one of them. I dont want it to flicker(or show another random iron 5km away) based on which detector is barely closer.
> 
> A possible compromise is to use the player as the center of the search with range of (detector range - detector distance).

Well, they have to be on same distance from player and have same range. while second is okay, the first is almost never gonna happen, unless done specially. It's a veeeery edge situation.

Racher commented on Data/Scripts/OreDetectorReforged/Detector/SearchTask.cs:
```diff
@@ -7,23 +7,26 @@ namespace OreDetectorReforged.Detector
         public List<IDetectorPage> pages;
 
-        public SearchTask(BoundingSphereD area, string minedOre, Func<Vector3D, bool> resultCb, Action finishCb)
+        public SearchTask(BoundingSphereD area, Vector3D center, string minedOre, Func<Vector3D, bool> resultCb, Action finishCb)
```
This confuses the principal function is no longer a nearest neighbor search. Instead, you start searching around center and if the nearest is not inside area you just break and find nothing. And if you didn't break it would be slow.

DeadMor0z replied:
I didn't change function itself. center is used to calculate distance from marker. area.center is still used for searching.

DeadMor0z replied:
I got it. These changes are irrelevant then) I misread how GPS markers work.

Racher commented on Data/Scripts/OreDetectorReforged/PlayerOreDetectorSession.cs:
```diff
@@ -39,7 +77,7 @@ public override void UpdateAfterSimulation()
             var ore = MaterialMappingHelper.Static.naturalOres[o];
             var area = new BoundingSphereD(detector.GetPosition(), settings.range);
-            DetectorServer.Add(new SearchTask(area, ore, settings.count, (results) =>
+            DetectorServer.Add(new SearchTask(area, MyAPIGateway.Session.Player.Character.GetPosition(), ore, settings.count, (results) =>
```
The player position is not always relevant

DeadMor0z replied:
what do you mean by that?

Racher replied:
I mean for eaxmple remote controlling a drone. But it's minor detail dont worry about it.

DeadMor0z replied:
Session.ControlledObject.Entity.GetPosition() - this will work for player and any controlled object

DeadMor0z replied:
Can I make another PullRequest? just discard this one.

Racher commented:
I made my attempt at the feature:
https://github.com/Racher/se-ore-detector-reforged/tree/broadcast
https://steamcommunity.com/sharedfiles/filedetails/?id=2807946569
Would you mind checking either the code or in-game to see if it covers what you wanted?

Commit:
Broadcasting: SearchTask parameters undo
Broadcasting: Using ControlledObject postiotn instead of Player position - fix distance in case of remote contorlling

DeadMor0z commented:
> I made my attempt at the feature: https://github.com/Racher/se-ore-detector-reforged/tree/broadcast https://steamcommunity.com/sharedfiles/filedetails/?id=2807946569 Would you mind checking either the code or in-game to see if it covers what you wanted?

Looks good.
